Step1. 엔티티 매핑

xonme888 opened the pull request:
2021-11-12 
- 좀 더 꼼꼼하게 공부하고 생각했다면 좋았을텐데 놓치고 있는 점들이 참 많았네요 다시 리뷰 드립니다. 😀







-------------
안녕하세요 
1단계 - 엔티티 매핑 리뷰 요청드립니다.

## 궁금한점 
1. entity 에서 `columnDefinition` 을 사용하게 되었는데요.   
- 자료형을 상수로 구현해서 사용하고자했습니다.
- 공통으로 사용하면 좋겠다 `DtaaBaseStructureConstant` 클래스를 만들어서 `public`으로 변경하려고보니 
- 유지보수 측면에서 varchar(20) 을 title과 name에서 쓰고 있는데 어느 한쪽만 변경 요구사항이 추가되었을 때  
- domain에서 변경이 아닌 상수를 수정하는경우 문제가 된다는 생각이들었습니다.
   - 커밋메시지 :  #feat(User) : 데이터베이스 자료구조 상수 추가
- 보편적으로 도메인 안에서 상수를 만들어서 쓰는지 아니면 다른 방법이 있는지 궁금합니다.

xonme888 committed:
docs(README) : 요구사항 정의 및 주요기능 이해 추가

xonme888 committed:
feat(application) : 설정추가

- sql를 보이게 설정
- h2에서 mysql 쿼리로 설정

xonme888 committed:
feat(BaseTime) : 사용자의 생성일시,수정일시 공통 추가

xonme888 committed:
feat(domain) : Create Entity

xonme888 committed:
test(AnswerTest) : Answer 테스트 추가

xonme888 committed:
test(QuestionTest) : Question 테스트 추가

xonme888 committed:
test(QuestionTest) : Question 테스트 추가

xonme888 committed:
test(UserRepository) : 유저Id로 유저 정보 조회 추가

xonme888 committed:
test(UserTest) : 유저 테스트 추가

xonme888 committed:
test(QuestionTest) : Id로 삭제되지 않은 건수인지 체크 추가

xonme888 committed:
docs : 주요기능 이해 추가

xonme888 committed:
feat(User) : 데이터베이스 자료구조 상수 추가

kang-hyungu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -5,13 +5,36 @@
+    @Column(name = "writer_id")
+    private Long writerId;
+
+    public Answer() {
```
접근제어자를 public으로 만들어야 하는 이유가 있을까요?
public 대신 다른 접근제어자를 쓸 수 없을까요?

kang-hyungu replied:
다른 도메인 객체의 생성자도 같이 점검해주세요

kang-hyungu commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,12 +1,38 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
+    @Column(columnDefinition = "varchar(100)")
```
columnDefinition 제거해주세요.
column 어노테이션의 다른 속성을 사용해서 길이제한을 만들 수 있습니다.
타입 명시는 따로 안해주셔도 됩니다. String이면 jpa가 알아서 varchar로 만들어줍니다.

kang-hyungu replied:
다른 객체도 같이 점검해보아요

kang-hyungu commented on src/main/java/qna/domain/UserRepository.java:
```diff
@@ -4,6 +4,10 @@
 public interface UserRepository extends JpaRepository<User, Long> {
     Optional<User> findByUserId(String userId);
+
+	Optional<User> findByName(String userId);
```
1. 들여쓰기가 윗줄이랑 다르게 보이네요
2. 파라미터명이 적절할까요?

kang-hyungu commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,66 @@
+	public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
+
+	@Autowired
+	private QuestionRepository questionRepository;
```
레파지토리 테스트는 따로 분리해주세요
QuestionTest면 Question 객체에 대한 테스트만 진행하도록 만들어주세요.

kang-hyungu reviewed (changes requested):
안녕하세요 승후님
몇 가지 간단한 피드백 남겼습니다.
수정해보시고 리뷰 재요청주세요

kang-hyungu commented:
>     1. entity 에서 `columnDefinition` 을 사용하게 되었는데요.
> 
> 
>     * 자료형을 상수로 구현해서 사용하고자했습니다.
> 
>     * 공통으로 사용하면 좋겠다 `DtaaBaseStructureConstant` 클래스를 만들어서 `public`으로 변경하려고보니
> 
>     * 유지보수 측면에서 varchar(20) 을 title과 name에서 쓰고 있는데 어느 한쪽만 변경 요구사항이 추가되었을 때
> 
>     * domain에서 변경이 아닌 상수를 수정하는경우 문제가 된다는 생각이들었습니다.
>       
>       * 커밋메시지 :  #feat(User) : 데이터베이스 자료구조 상수 추가
> 
>     * 보편적으로 도메인 안에서 상수를 만들어서 쓰는지 아니면 다른 방법이 있는지 궁금합니다.

1. 필요하다면 도메인 객체에 상수를 정의하고 사용해도 문제 없습니다.
2. 작성해주신 케이스에서는 상수로 정의하지 않는걸 추천드립니다. 상수로 정의해서 같이 사용하다가 변경사항이 생겨서 title을 100자로 바꾸게 되면 다른 스키마도 같이 바뀌는 문제가 발생하겠네요. 이 경우에는 각각 정의하는게 안전합니다.

xonme888 committed:
fix(Entity) : columnDefinition 삭제 후 length 수정

xonme888 committed:
fix(Entity) : repository와 Entity클래스 테스트 분리

xonme888 committed:
fix(Entity) : 파라미터명 변경

xonme888 committed:
fix(Entity) : @DataJpaTest 추가

xonme888 committed:
feat(User) : 유저 이름 찾는 테스트 추가 및 중복 이름 수정

xonme888 committed:
test(User) : 객체의 동시성과 동일하지 않음을 확인

xonme888 committed:
fix(User) : 유저의 id 유니크 설정

xonme888 committed:
fix(DeleteHistory) : 생성시만 저장되도록 수정 안되도록 추가

xonme888 committed:
docs(README) 공부한 내용정리

xonme888 committed:
test(Answer) : Answer 테스트 추가

QuestionTest 추가 :
- 테스트를 편하게 하기 위해 스태틱 데이터에 Pk 값 추가

xonme888 committed:
feat(ErrorMessage) : 공통 예외 메시지 정의

xonme888 committed:
test(Question) : 테스트 추가

xonme888 committed:
test(UserTest) : 유저테스트 추가

xonme888 committed:
test(Answer) : AnswerRepositoryTest 수정

xonme888 committed:
test(QuestionRepository) : 테스트추가

xonme888 committed:
test(UserRepositoryTest) : 테스트추가

xonme888 committed:
docs(README) qna 분석

xonme888 committed:
style: 들여쓰기 변수명수정

xonme888 committed:
test(User): 유저 테스트 추가 및 상수 메시지 추가

- 유저 권한 메시지 추가
- 유저 비밀번호를 변경시 변경되는지 확인
- 동일한 유저의 정보가 아닌경우 수정시 예외
- 동일한 유저의 계정을 수정시 변경되었는지 확인

xonme888 committed:
test(DeleteHistoryTest): 서로 다른 정보를 넣는 경우 다른 데이터인지 확인

xonme888 committed:
feat(Question): 객체의 동일성 보장을 위해 equals 재정의

xonme888 committed:
test(Answer): 컬렉션에 담아서 저장, 동일한 객체를 여러번 저장 테스트

xonme888 committed:
test(DeleteHistoryTest): ContentType_Null이면_예외 테스트 추가

xonme888 committed:
feat(DeleteHistory): 파라미터 정보가 null 인경우 예외처리

xonme888 committed:
test(DeleteHistoryTest): 파라미터 정보가 null 인경우 예외 검증

xonme888 committed:
docs(README): 이해한점 추가

xonme888 committed:
fix(README): 데이터 동일성 검증 수정

xonme888 committed:
test(README) : autoincrement 리셋 로직 추가

xonme888 committed:
feat(Application) : 강의로 인한 패키지 탐색 추가

xonme888 committed:
feat(application) : 설정 삭제

xonme888 committed:
feat(build.gradle) : 강의로 이한 테스트 범위 exclude

xonme888 committed:
docs(README) : 무엇을 알게되었는지 추가

xonme888 committed:
test(AnswerRepositoryTest) : 여러건을_저장할때_동일한_객체라면_동일한객체는_변경이_없다면_INSERT되지_않는다

kang-hyungu commented on src/main/java/qna/domain/ContentType.java:
```diff
@@ -1,5 +1,18 @@
+import static java.util.stream.Collectors.*;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
```
사용하지 않는 import는 정리해주세요

kang-hyungu commented on src/main/java/qna/domain/ContentType.java:
```diff
@@ -1,5 +1,18 @@
+    QUESTION, ANSWER;
+
+
+    public static String valuesString(){
```
이 메서드는 toString 때문에 추가하신 걸까요?
이 메서드가 필요한가요?

kang-hyungu replied:
예외 메시지와 테스트에서 사용하고 계시군요.
이 형태로 보여줄 필요가 있는지 모르겠네요.
contentType은 필수 입력입니다 와 같은 메시지를 뿌리면 되지 않을까요?

xonme888 replied:
아! 최근 요구사항에 타입을 명시해달라는 말 때문에 저렇게 처리하게 되었네요 ! 😭 
수정해보겠습니다.

kang-hyungu commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+	void 여러건을_저장할때_동일한_객체라면_동일한객체는_변경이_없다면_INSERT되지_않는다() {
+
+		Answer save = answerRepository.save(A1);
+		System.out.println("save = " + save);
```
println 구문은 제거해주세요.

kang-hyungu commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+
+@DataJpaTest
+public class AnswerRepositoryTest {
```
테스트 코드는 들여쓰기를 탭으로 사용하셨네요
전체 소스코드의 들여쓰기를 통일시켜주세요.

kang-hyungu commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	@AfterEach
+	public void teardown() {
```
이 메서드도 제거해주세요.

kang-hyungu commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+
+	@Test
+	@DisplayName("기대값을 3건을 예상했지만 결과는 2건이다.")
+	void 여러건을_저장할때_동일한_객체라면_동일한객체는_변경이_없다면_INSERT되지_않는다() {
```
왜 3건이 아닌 2건으로 나왔을까요?
teardown 메서드를 없애고 이 테스트 메서드만 실행하면 성공하는데 전체 테스트를 돌리면 실패하는 이유가 무엇일까요?

힌트
teardown 메서드를 주석처리하고 이 테스트 메서드만 실행 했을 때와 전체 테스트를 각각 돌려보면서 A1 객체가 어떻게 다른지 살펴봅니다.
그리고 hibernate 로그도 어떻게 다른지 비교해보시기 바랍니다.

kang-hyungu reviewed (approved):
승후님 피드백 반영 잘하셨습니다.
추가로 남긴 코멘트는 다음 미션에서 반영해주세요
pr은 머지 처리하겠습니다. 수고하셨어요
